src/serializers/graphcontext_defrag.c

```diff
@@ -0,0 +1,306 @@
+	SIValue *v
+) {
+	SIType t = SI_TYPE (*v) ;
+
+	// skip non-heap values quickly
+	if (!SI_HEAP_ALLOCATED (*v)) {
+		return ;
+	}
+
+	void *p = NULL ;
+	void **ref_p = NULL ;
+
+	switch (t) {
+		case T_ARRAY:
+			defrag_array (ctx, v) ;
+			return ;
+
+		case T_VECTOR_F32:
+			p = v->ptrval ;
+			ref_p = &v->ptrval ;
+			break ;
+
+		case T_INTERN_STRING:
+			// do not defrag interned strings, as these are managed
+			// by a string pool
+			return ;
+
+		case T_STRING:
+			p     = v->stringval ;
+			ref_p = (void**)&v->stringval ;
+			break ;
+
+		default :
+			ASSERT (false && "unexpected value type") ;
+			break ;
+	}
+
+	ASSERT (p     != NULL) ;
+	ASSERT (ref_p != NULL) ;
+
+	void *moved = RedisModule_DefragAlloc (ctx, p) ;
+	if (moved != NULL) {
+		*ref_p = moved ;
+	}
+}
+
+static void defrag_array
+(
+	RedisModuleDefragCtx *ctx,
+	SIValue *arr
+) {
+	SIValue _arr = *arr ;
+	if (!SIArray_ContainsType (_arr, SI_HEAP)) {
+		return ;
+	}
+
+	u_int32_t l = SIArray_Length (_arr) ;
+	for (u_int32_t i = 0; i < l; i++) {
+		SIValue *v = SIArray_GetRef (_arr, i) ;
+		defrag_value (ctx, v) ;
+	}
+}
+
+// defrag an AttributeSet
+// ensure any heap pointers are moved using RedisModule_DefragAlloc
+static void defrag_attributeset
+(
+	RedisModuleDefragCtx *ctx,
+	AttributeSet *set
+) {
+	void *moved = NULL ;
+	AttributeSet _set = *set ;
+	uint16_t n = AttributeSet_Count (_set) ;
+
+	// defrag set
+	moved = RedisModule_DefragAlloc (ctx, _set) ;
+	if (moved != NULL) {
+		*set = moved ;
+		_set = *set  ;
+	}
+
+	// defrag attributes
+	for (uint16_t i = 0; i < n ; i++) {
+		// read current SIValue
+		SIValue *v = AttributeSet_GetIdxRef (_set, i, NULL) ;
+		defrag_value (ctx, v) ;
+	}
+}
+
+// defrag entities (both nodes and edges)
+static int defrag_entities
+(
+	RedisModuleDefragCtx *ctx,
+	defrag_stage stage,
+	const Graph *g,
+	GraphContext *gc,
+	DataBlockIterator *it
+) {
+	uint64_t counter = 0 ;
+	AttributeSet *set = NULL ;
+
+	// get current entity attribute-set
+	while ((set = (AttributeSet*)(DataBlockIterator_Next (it, NULL))) != NULL) {
+		counter++ ;
+
+		// entity has no attributes, skip
+		if (*set == NULL) {
+			continue ;
+		}
+
+		defrag_attributeset (ctx, set) ;
+
+		// check if we should stop
+        if ((counter % 64 == 0) && RedisModule_DefragShouldStop (ctx)) {
+			// only pause if NOT at the end
+			if (!DataBlockIterator_Depleted (it)) {
+				// save current stage and offset
+				_save_stage (ctx, stage, DataBlockIterator_Position (it)) ;
+
+				return 1;
+			}
+			// else: fall through, loop will terminate, return 0
+        }
+	}
+
+	// iterator exhausted, no more work
+	return 0 ;
+}
+
+static int defrag_edges
+(
+	RedisModuleDefragCtx *ctx,
+	GraphContext *gc,
+	uint64_t offset
+) {
+	Graph *g = GraphContext_GetGraph (gc) ;
+	DataBlockIterator *it = Graph_ScanEdges (g) ;
+	DataBlockIterator_Seek (it, offset) ;  // seek iterator to offset
+
+	// obtain exclusive access to the graph
+	Graph_AcquireWriteLock (g) ;
```

Comment:
See if we can introduce a timeout for this operation? e.g. fail locking if we didn't acquire after specified time.